commit b4ea02ea2899c35d69b729937925cf2825067fd0 (refs/changes/13/5313/2)
Author: Robert Bogart <Robert.bogart@joyent.com>
Date:   2019-01-04T18:41:27+00:00 (9 months ago)
    
    MANATEE-343 manatee full rebuilds should pull from the primary where possible

diff --git a/lib/manatee-peer.js b/lib/manatee-peer.js
index b6d242f..d343a30 100644
--- a/lib/manatee-peer.js
+++ b/lib/manatee-peer.js
@@ -976,34 +976,41 @@ ManateePeer.prototype.startInitialSetup = function ()
 	});
 };
 
-/*
- * Return the desired postgres configuration for the current cluster state.
- */
-ManateePeer.prototype.pgConfig = function ()
+ManateePeer.prototype.pgCfgInit = function (role, upstream, downstream, primary)
 {
 	var config = {};
 
-	if (this.mp_role == MP_ROLE_PRIMARY) {
-		config.role = 'primary';
-		config.upstream = null;
-		config.downstream = this.mp_zkstate.sync;
-	} else if (this.mp_role == MP_ROLE_SYNC ||
-	    this.mp_role == MP_ROLE_ASYNC) {
-		config.role = this.mp_role == MP_ROLE_SYNC ?
-		    'sync' : 'async';
-		config.upstream = this.mp_pg_upstream;
-		config.downstream = null;
-	} else {
-		mod_assertplus.ok(this.mp_role == MP_ROLE_UNASSIGNED ||
-		    this.mp_role == MP_ROLE_DEPOSED);
-		config.role = 'none';
-		config.upstream = null;
-		config.downstream = null;
-	}
+	config.role = role;
+	config.upstream = upstream;
+	config.downstream = downstream;
+	config.primary = primary;
 
 	return (config);
 };
 
+/*
+ * Return the desired postgres configuration for the current cluster state.
+ */
+ManateePeer.prototype.pgConfig = function (role, upstream, downstream, primary)
+{
+	var config;
+
+        if (this.mp_role == MP_ROLE_PRIMARY) {
+                config = this.pgCfgInit('primary', null, this.mp_zkstate.sync,
+                    null);
+        } else if (this.mp_role == MP_ROLE_SYNC ||
+            this.mp_role == MP_ROLE_ASYNC) {
+                var role = this.mp_role == MP_ROLE_SYNC ? 'sync' : 'async';
+                config = this.pgCfgInit(role, this.mp_pg_upstream, null,
+                    this.mp_zkstate.primary);
+        } else {
+                mod_assertplus.ok(this.mp_role == MP_ROLE_UNASSIGNED ||
+                    this.mp_role == MP_ROLE_DEPOSED);
+                config = this.pgCfgInit('none', null, null, null);
+        }
+	return (config);
+};
+
 /*
  * Reconfigure postgres based on the current configuration.  During
  * reconfiguration, new requests to reconfigure will be ignored, and incoming
